### ISSUE TITLE
Inconsistent gradient results in F.grid_sample using torch.autograd.grad with create_graph=True

### ISSUE THREAD

**Issue by nikcheerla** [pytorch/pytorch#8080]
## Issue description

I am trying to work with gradients using the `F.grid_sample` function from STNs. The issue I have is that sometimes the output of `torch.autograd.grad()` is a tensor with `requires_grad() = True`, and sometimes it has `requires_grad() = False` instead (based on what seem to be trivial modifications to the code). I assumed that when using `create_graph=True`, the output would always be a differentiable tensor? Was this assumption incorrect? Could this be a bug in `F.grid_sample()` or in `autograd`, or am I somehow implementing this wrong?

## Code example

I've provided a minimal example of the issue here.

```python
import torch
import torch.nn as nn
import torch.nn.functional as F

a = torch.randn(1, 3, 32, 32).requires_grad_().float()
b = torch.tensor(0.01).requires_grad_().float()

def f(a, b):
	grid = F.affine_grid(torch.eye(3).unsqueeze(0)[:, 0:2], \
				size=torch.Size((1, 2, 64, 64)))
	return F.grid_sample(a+b, grid).mean()

f_val = f(a, b)

(dfda,) = torch.autograd.grad(f_val, [a], create_graph=True, retain_graph=True)
print (f"df/da={dfda.shape}, requires_grad={dfda.requires_grad}")

def g(a, b):
	grid = F.affine_grid(torch.eye(3).unsqueeze(0)[:, 0:2], \
				size=torch.Size((1, 2, 64, 64)))

	# Basically the same function, just with multiplication instead of addition
	return F.grid_sample(a*b, grid).mean() 

g_val = g(a, b)

(dgda,) = torch.autograd.grad(g_val, [a], create_graph=True, retain_graph=True)
print (f"dg/da={dgda.shape}, requires_grad={dgda.requires_grad}")
```
Output:
```
df/da=torch.Size([1, 3, 32, 32]), requires_grad=False
dg/da=torch.Size([1, 3, 32, 32]), requires_grad=True
```
The functions `f()` and `g()` are almost identical, except one has multiplication instead of addition, yet the outputs differ in the requires_grad property.

## System Info
PyTorch version: 0.4.0
Is debug build: No
CUDA used to build PyTorch: None

OS: Mac OSX 10.13.2
GCC version: Could not collect
CMake version: version 3.5.2

Python version: 3.6
Is CUDA available: No
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA

Versions of relevant libraries:
[pip] numpy (1.13.3)
[pip] numpydoc (0.6.0)
[pip] torch (0.4.0)
[pip] torchvision (0.2.1)
[conda] pytorch                   0.4.0           py36_cuda0.0_cudnn0.0_1    pytorch
[conda] torchvision               0.2.1                    py36_1    pytorch

**Comment by bhushan23** [comment]
>>Could this be a bug in F.grid_sample() or in autograd, or am I somehow implementing this wrong?
I think, you are implementing correctly.  Issue is mostly in grid_sample() or in autograd

**Comment by bhushan23** [comment]
I have confirmed that the problem is with torch.autograd.grad 
both f_val and g_val does have requires_grad True
Debugging into grad to see what's going on

**Comment by bhushan23** [comment]
@soumith I think we can drop this issue as gradient of a+b wrt to a will be 1 and we don't need to set requires_grad for this. Correct me if I am wrong.

**Comment by soumith** [comment]
I confirmed with @apaszke as well that this is expected behavior.

He says:

because create_graph=True, this is expected.
In the a+b case, basically the gradient of the gradient (wrt parameters) is all 0, which means it doesn’t require grad, because it effectively never touched a single variable that does require grad.

**Comment by soumith** [comment]
thanks for looking into it @bhushan23
